### ISSUE TITLE
[Bug] [Low priority] Fee rates differ from set rate.

### ISSUE THREAD

**Issue by junderw** [lightningnetwork/lnd#620]
VSize 145 transaction was sent with --sat_per_byte=500 and the resulting transaction was 606.03 sat/B.

dividing fee by full size, or even size of legacy-block-tx does not go anywhere near 500 sat/B.

I am guessing that the byte size estimation logic is broken but haven't looked into it.

The only variable part of transactions is the size of r and s values of the signatures. (r can be 33 or 32 bytes 50:50 and s can be 32 or 31 bytes 99.6:0.4) but this transaction was one input one signature... so a 100 sat/b variance is not caused by that.

**Comment by MeshCollider** [comment]
Is this using `sendcoins` RPC? And the log shows the correct sat/byte like `[sendcoins] addr=X, amt=X, sat/byte=x`? I believe the actual usage of the sat/byte is just passed on to btcwallet, it isn't actually used by lnd, so perhaps the bug lies upstream

**Comment by junderw** [comment]
nope. `lncli openchannel`

**Comment by junderw** [comment]
Update: I used sendcoins and the estimate was correct(ish) for actual raw bytes (every byte counted as 1)

Size 341
VSize 179
Fee 8148

Sat/byte 23.894428152
Sat/Vbyte 45.519553073

My command:
`$ lncli sendcoins --sat_per_byte=20 aaaaaaaaa xxxxxx`

So it looks like sendcoins is looking at size and not Vsize (discounted size) as 23.89 seems like a reasonable margin of error... I'm sure it was trying to get rid of dust as well since I had to eyeball the send amount (I lowered it 1000 sats at a time until it sent) and if you subtract 1000 sats from the fee, fee/size will be around 20 sats/byte.

Conclusion: **sendcoins is not considering segwit for fee calculations**

Still unknown: **openchannel** seems way off from any estimate... I wonder if there's something wrong with size calculations for the opening transaction of a channel...

**Comment by bretton** [comment]
Dropping this here as I believe it's the same issue. lncli openchannel has strangely high fees
```
0.001      channel, 0.00004218 BTC fees (manual)
0.01       channel, 0.00088595 BTC fees (manual)
0.16777216 channel, 0.00001222 BTC fees (autopilot)
0.16777216 channel, 0.00001222 BTC fees (autopilot)
0.16777216 channel, 0.00000612 BTC fees (autopilot)
```

**Comment by Roasbeef** [comment]
>  lncli openchannel has strangely high fees

Strangely high compared to what? It's using testnet fee estimation which is generally pretty wacky. 

As for `lncli sendcoins`, the internal logic within `btcwallet` is indeed _not_ aware of witness outputs/outputs, etc. Atm, it treats them all as if they were regular `p2pkh`.

**Comment by junderw** [comment]
@Roasbeef FYI, all issues I am experiencing in this issue are on mainnet, so the "testnet is wonky" argument doesn't apply.

(Yes, I know, test on testnet, but fee related issues like this are hard to report on testnet, as you say, because testnet fees are wonky to begin with)

**Comment by Roasbeef** [comment]
@junderw my comment wasn't directed at your initial report. See my the second paragraph in my last response about the difference with `sendcoins`. 

As for `openchannel`, all our estimates are based on weight rather than bytes. We mechanically convert from sat/bytes to sat/weight, but this itself is lossy. The solution is to either return weight directly from fee estimates, switch `openchannel` estimation to use bytes, or just switch over to `vbytes` internally uniformly for fee estimation. The last one there seems like the "most correct" solution.

**Comment by 2pac1** [comment]
There might be an issue:

lncli openchannel 02669e1e43577c14a4ae796a38e0cfdadb548615058e367bd96bf825e62c0932ce 100000 --sat_per_byte=5
{
	"funding_txid": "e0b1bb3ff2db5417a297138e7e95d62cec4dec19ee32729228823cc2112c0d83"
}
https://www.smartbit.com.au/tx/e0b1bb3ff2db5417a297138e7e95d62cec4dec19ee32729228823cc2112c0d83
I have set it to 5 sat/byte, and it is showing 3.99


Tested it again:

 lncli openchannel 03cbf298b068300be33f06c947b9d3f00a0f0e8089da3233f5db37e81d3a596fe1 100000 50000 --sat_per_byte=10
{
	"funding_txid": "e267e54872053a7618567f31a9d27e38cdbff0e417614466f39df4d2513704bf"
}

Set it to 10, and it showing 7.99
https://www.smartbit.com.au/tx/e267e54872053a7618567f31a9d27e38cdbff0e417614466f39df4d2513704bf

Is this normal?
Tnx guys
(this is mainnet)

**Comment by halseth** [comment]
@2pac1 Yeah, it's a bit buggy. 

I'm currently looking into this!

**Comment by 2pac1** [comment]
ok cool
let me know whats going on

**Comment by Roasbeef** [comment]
Fixed by #762, https://github.com/Roasbeef/btcwallet/pull/16 and two commits that landed into my fork of `btcd.
